source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,102 @@
+---
+Title: Engine Dedicated CPUs
+Category: feature
+Authors: sabusale
+---
+
+# Engine Dedicated CPUs
+
+## Summary
+
+Once choosing dedicated CPU policy, The guest vCPUs will be strictly pinned to a set of host pCPUs
+(similarly to static CPU pinning). The set of pCPUs will be chosen to match the required guest CPU topology.
+
+Engine handles the pinning automatically when starting the VM as opposed to the manual pinning that is
+done by the user and is static.
+
+## Owner
+
+-   Name: Saif Abu Saleh (sabusale)
+-   Email: sabusale@redhat.com
+
+## Responsibilities
+
+VDSM:
+
+- Manages the available pinned CPUs
+
+Engine:
+- Schedule VM to run on best satisifing host
+- creates vCpu to pCpu mappings
+
+
+## Backend
+
+1. Add dedicated to CPU Pinning Policy enum
+
+2.  Scheduling
+    1. Start VM:
+        - Scheduling phase
+            - Score all hosts to find best matching host with enough pCPUs
+        - Run phase
+            - Create vCpu to pCpu mapping based on the host free pCPUs reported
+            - Update VM CurrentCpuPinning DB field with the mapping
+
+    2. Stop/Destroy VM:
+        - Clear VM dynamic CPU pinning field
+    
+3. Send vcpupin with cpusets and policy name in VM metadata
+   1. Send the policy name, available policy names: { none, pin, dedicated, siblings, isolate-threads }.
+   2. Send the cpusets (it will be stored in VM_Dynamic)
+
+4. Migrate VDSM call
+   1. Calculate vCPU to pCPU mapping based on the scheduled host
+   2. Send cpusets additional parameter
+   
+5. setNumberOfCpus VDSM call
+   1. Send cpusets additional parameter
```

Comment:
of all CPUs, not just the plugged ones
need to see what to do with unplugged ones - I guess we unplug the vCPUs with the highest indexes but I'm not sure

Reply:
This can be more complicated than it seems:

- what if the VM is dedicated and added CPU's doesn't fit based on the policy?

- un-plug with the new policies, we need to see what to do

optimally we would want to check where there is cores that is used by that VM that we can free, so for example we don't hold core with 1 CPU dedicated (in this case other VMs is blocked for siblings/isolate threads policy on that core)

(maybe same for sockets?)

Reply:
> This can be more complicated than it seems:
> 
>     * what if the VM is dedicated and added CPU's doesn't fit based on the policy?

so we shouldn't plug I guess

> 
>     * un-plug with the new policies, we need to see what to do
> 
> 
> optimally we would want to check where there is cores that is used by that VM that we can free, for example if we have 1 CPU dedicated on the whole core (in this case other VMs is blocked for siblings/isolate threads policy on that core) so we should free it if possible

yeah, but that's introduces extra complexity (also need to change the engine<->vdsm API) - doesn't seem to worth it